types/vote_test.go

```diff
@@ -490,3 +490,67 @@ func TestVoteProtobuf(t *testing.T) {
 		}
 	}
 }
+
+func TestSignAndCheckVote(t *testing.T) {
+	height, round := int64(1), int32(0)
+	privVal := NewMockPV()
+	pk, err := privVal.GetPubKey()
+	require.NoError(t, err)
+
+	testCases := []struct {
+		name              string
+		extensionsEnabled bool
+		vote              *Vote
+		expectError       bool
+	}{
+		{
+			name:              "precommit with extension signature",
+			extensionsEnabled: true,
+			vote:              examplePrecommit(),
+			expectError:       false,
+		},
+		{
+			name:              "precommit with extension signature for a nil block",
+			extensionsEnabled: true,
+			vote: &Vote{
+				ValidatorAddress:   pk.Address(),
+				ValidatorIndex:     0,
+				Height:             height,
+				Round:              round,
+				Timestamp:          cmttime.Now(),
+				Type:               PrecommitType,
+				BlockID:            BlockID{make([]byte, 0), PartSetHeader{0, make([]byte, 0)}},
+				ExtensionSignature: []byte("signature"),
+			},
+			expectError: false,
+		},
+		{
+			name:              "precommit with extension signature for a nil block, but extensionsEnabled is false",
+			extensionsEnabled: false,
+			vote: &Vote{
+				ValidatorAddress:   pk.Address(),
+				ValidatorIndex:     0,
+				Height:             height,
+				Round:              round,
+				Timestamp:          cmttime.Now(),
+				Type:               PrecommitType,
+				BlockID:            BlockID{make([]byte, 0), PartSetHeader{0, make([]byte, 0)}},
+				ExtensionSignature: []byte("signature"),
+			},
+			expectError: false,
+		},
```

Comment:
This code is not test:

if !isPrecommit && extensionsEnabled

Reply:
This code is also not tested:

if extSignature == (!isPrecommit || isNil) {

Reply:
I think it is more complex, but can we also test when signing a vote fails? Namely:

if err := privVal.SignVote(chainID, v); err != nil

Reply:
done